compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -397,6 +406,170 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
+                dlog.error(bLangTypeTwo.pos, INVALID_ERROR_INTERSECTION, bLangTypeTwo);
+                continue;
+            }
+
+            // Collecting typeIds.
+            if (typeOne.tag != TypeTags.READONLY) {
+                populateSecondaryTypeIdSet(secondaryTypeIds, (BErrorType) typeOne);
+            }
+            if (typeTwo.tag != TypeTags.READONLY) {
+                populateSecondaryTypeIdSet(secondaryTypeIds, (BErrorType) typeTwo);
+            }
+
+            BType potentialIntersectionType = types.getTypeIntersection(typeOne, typeTwo);
+
+            boolean isValidIntersection = true;
+            for (int i = 2; i < constituentTypes.size(); i++) {
+                BLangType bLangType = constituentTypes.get(i);
+                BType bType = symResolver.resolveTypeNode(bLangType, env);
+
+                if (bType.getKind() != TypeKind.ERROR && typeTwo.tag != TypeTags.READONLY) {
+                    dlog.error(bLangType.pos, INVALID_ERROR_INTERSECTION, bLangType);
+                    isValidIntersection = false;
+                    break;
+                }
+
+                if (bType.tag != TypeTags.READONLY) {
+                    populateSecondaryTypeIdSet(secondaryTypeIds, (BErrorType) bType);
+                }
+
+                potentialIntersectionType = types.getTypeIntersection(potentialIntersectionType, bType);
+                if (potentialIntersectionType == symTable.semanticError) {
+                    dlog.error(typeDefinition.typeNode.pos, INVALID_INTERSECTION_TYPE, typeDefinition.typeNode);
+                    isValidIntersection = false;
+                    break;
+                }
+            }
+
+            if (!isValidIntersection) {
+                continue;
+            }
+
+            if (isDistinctType) {
+                ((BErrorType) potentialIntersectionType).typeIdSet = BTypeIdSet.from(
+                        potentialIntersectionType.tsymbol.pkgID, typeDefinition.name.value, isPublicType,
+                        secondaryTypeIds);
+            }
+
+            LinkedHashSet<BType> constituentBTypes = new LinkedHashSet<>() {{
+                add(typeOne);
+                add(typeTwo);
+            }};
+            BIntersectionType intersectionType = definePotentialIntersectionType((BErrorType) potentialIntersectionType,
+                                                                                 typeDefinition, constituentBTypes);
+            typeDefinition.symbol = intersectionType.tsymbol;
+            typeDefinition.typeNode.type = intersectionType;
+        }
+    }
+
+    private void populateSecondaryTypeIdSet(Set<BTypeIdSet.BTypeId> secondaryTypeIds, BErrorType typeOne) {
+        secondaryTypeIds.addAll(typeOne.typeIdSet.primary);
+        secondaryTypeIds.addAll(typeOne.typeIdSet.secondary);
+    }
+
+    private BIntersectionType definePotentialIntersectionType(BErrorType potentialIntersectionType,
```

Comment:
In this case, is it actually potential if we get to this point?

Reply:
fixed